framework/aster-frame/src/vm/page_table/properties.rs

```diff
@@ -0,0 +1,247 @@
+
+    /// The number of bits used to index a PTE in a page table frame.
+    const IN_FRAME_INDEX_BITS: usize = Self::NR_ENTRIES_PER_FRAME.ilog2() as usize;
+
+    /// The index of a VA's PTE in a page table frame at the given level.
+    fn in_frame_index(va: Vaddr, level: usize) -> usize {
+        va >> (Self::BASE_PAGE_SIZE.ilog2() as usize + Self::IN_FRAME_INDEX_BITS * (level - 1))
+            & (Self::NR_ENTRIES_PER_FRAME - 1)
+    }
+
+    /// The page size at a given level.
+    fn page_size(level: usize) -> usize {
+        Self::BASE_PAGE_SIZE << (Self::IN_FRAME_INDEX_BITS * (level - 1))
+    }
+}
+
+bitflags::bitflags! {
+    /// The status of a memory mapping recorded by the hardware.
+    pub struct MapStatus: u8 {
+        const ACCESSED = 0b0000_0001;
+        const DIRTY    = 0b0000_0010;
+    }
+}
+
+/// The cache policy of a memory mapping.
+/// FIXME: This may not be supported by all architectures and could be
+/// ignored by us without warnings at the moment.
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum MapCachePolicy {
+    Uncachable,
+    WriteCombining,
+    WriteThrough,
+    WriteBack,
+    WriteProtected,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct MapProperty {
+    pub perm: VmPerm,
+    /// Global.
+    /// A global page is not evicted from the TLB when TLB is flushed.
+    pub(crate) global: bool,
+    /// The properties of a memory mapping that is used and defined as flags in PTE
+    /// in specific architectures on an ad hoc basis. The logics provided by the
+    /// page table module will not be affected by this field.
+    pub(crate) extension: u64,
+    pub(crate) cache: MapCachePolicy,
+}
+
+/// Any functions that could be used to modify the map property of a memory mapping.
+///
+/// To protect a virtual address range, you can either directly use a `MapProperty` object,
+///
+/// ```rust
+/// let page_table = KERNEL_PAGE_TABLE.get().unwrap();
+/// let prop = MapProperty {
+///     perm: VmPerm::R,
+///     global: true,
+///     extension: 0,
+///     cache: MapCachePolicy::WriteBack,
+/// };
+/// page_table.protect(0..PAGE_SIZE, prop);
+/// ```
+///
+/// use a map operation
+///
+/// ```rust
+/// let page_table = KERNEL_PAGE_TABLE.get().unwrap();
+/// page_table.map(0..PAGE_SIZE, cache_policy_op(MapCachePolicy::WriteBack));
+/// page_table.map(0..PAGE_SIZE, perm_op(|perm| perm | VmPerm::R));
+/// ```
+///
+/// or even customize a map operation using a closure
+///
+/// ```rust
+/// let page_table = KERNEL_PAGE_TABLE.get().unwrap();
+/// page_table.map(0..PAGE_SIZE, |info| {
+///     assert!(info.prop.perm.contains(VmPerm::R));
+///     MapProperty {
+///         perm: info.prop.perm | VmPerm::W,
+///         global: info.prop.global,
+///         extension: info.prop.extension,
+///         cache: info.prop.cache,
+///     }
+/// });
+/// ```
+pub trait MapOp: Fn(MapInfo) -> MapProperty {}
+impl<F> MapOp for F where F: Fn(MapInfo) -> MapProperty {}
```

Comment:
This `MapOp` trait, the multiple `impl` blocks for `FnOnce<(MapInfo,)>`, and the helper functions (`cache_policy_op` and `perm_op`) look an unnecessarily complex solution to me.

I am particularly unconformable with the two functions of `cache_policy_op` and `perm_op`, which take a closure to return a new closure. This is too magical and implicit to my taste. The very design goal of mapping an old `MapProperty` to a new one in a concise way can be fulfilled in a more idiomatic and explicit way, without compromising the conciseness.

This can be simply done by changing the signature of a user function like `PageTable::protect` from the original form

```rust
    pub(crate) fn protect(
        &self,
        vaddr: &Range<Vaddr>,
        op: impl MapOp, // Fn(MapInfo) -> MapProperty
    ) -> Result<(), PageTableError> {
```

to the following form

```rust
    pub(crate) fn protect<F: Fn(&mut MapProperty)>(
        &self,
        vaddr: &Range<Vaddr>,
        fn: F,
    ) -> Result<(), PageTableError> {
```

Changing the signature does not seem to have any impact on the efficiency or correctness of the `protect` method.

But the caller side can be simplified. For example, an user using the original API looks as follows.

```rust
pt.protect(&prot, perm_op(|p| p - VmPerm::W))
```

Now with the new API, the user code is still simple enough, but does not need the magic `perm_op` utility function.

```rust
pt.protect(&prot, |old_prop| old_prop.perm -= VmPerm::W)
```